multisense_ros/src/multisense.cpp

```diff
@@ -0,0 +1,1651 @@
+        {
+            ros_image.encoding = sensor_msgs::image_encodings::TYPE_32FC1;
+            ros_image.step = image.width * 4;
+            break;
+        }
+        default: {throw std::runtime_error("Invalid image format");}
+    };
+
+    ros_image.is_bigendian = (htonl(1) == 1);
+}
+
+void publish_image(const lms::Image &image,
+                   std::shared_ptr<ImagePublisher> publisher,
+                   sensor_msgs::msg::Image &ros_image,
+                   const std::string &frame_id,
+                   bool use_ptp_time)
+{
+    populate_image(image, ros_image, frame_id, use_ptp_time);
+
+    auto camera_info = create_camera_info(image.calibration, ros_image.header, image.width, image.height);
+    publisher->publish(std::make_unique<sensor_msgs::msg::Image>(ros_image),
+                       std::make_unique<sensor_msgs::msg::CameraInfo>(std::move(camera_info)));
+}
+
+template <typename Color>
+void publish_point_cloud(const lms::PointCloud<Color> &point_cloud,
+                         const lms::TimeT &time,
+                         rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr publisher,
+                         sensor_msgs::msg::PointCloud2 &ros_point_cloud,
+                         const std::string &frame_id)
+{
+    ros_point_cloud.is_bigendian = (htonl(1) == 1);
+    ros_point_cloud.is_dense = true;
+    ros_point_cloud.header.frame_id = frame_id;
+    ros_point_cloud.fields.reserve(4);
+    ros_point_cloud.fields.resize(3);
+    ros_point_cloud.point_step = 0;
+    ros_point_cloud.fields[0].name     = "x";
+    ros_point_cloud.fields[0].offset   = 0;
+    ros_point_cloud.fields[0].count    = 1;
+    ros_point_cloud.fields[0].datatype = sensor_msgs::msg::PointField::FLOAT32;
+    ros_point_cloud.point_step += sizeof(float);
+    ros_point_cloud.fields[1].name     = "y";
+    ros_point_cloud.fields[1].offset   = sizeof(float);
+    ros_point_cloud.fields[1].count    = 1;
+    ros_point_cloud.fields[1].datatype = sensor_msgs::msg::PointField::FLOAT32;
+    ros_point_cloud.point_step += sizeof(float);
+    ros_point_cloud.fields[2].name     = "z";
+    ros_point_cloud.fields[2].offset   = 2 * sizeof(float);
+    ros_point_cloud.fields[2].count    = 1;
+    ros_point_cloud.fields[2].datatype = sensor_msgs::msg::PointField::FLOAT32;
+    ros_point_cloud.point_step += sizeof(float);
+
+    if constexpr (std::is_same_v<Color, std::array<uint8_t, 3>>)
+    {
+        ros_point_cloud.fields.resize(4);
+        ros_point_cloud.fields[3].name = "rgb";
+        ros_point_cloud.fields[3].offset = 3 * sizeof(float);
+        ros_point_cloud.fields[3].count = 1;
+        ros_point_cloud.fields[3].datatype = sensor_msgs::msg::PointField::FLOAT32;
+        ros_point_cloud.point_step += sizeof(float);
+
+        const size_t num_fields = ros_point_cloud.fields.size();
+
+        //
+        // Unfortunately we need to iterate and write our points to the output buffer since we need to
```

Comment:
not sure what the receiving end looks like, but just curious if it can use the field configuration to pull the color channels out of 3 separate uint8 fields instead of 1 float32?